### PR TITLE
Add `maybeResize` (and friends)

### LINKED ISSUE #2779: Add `maybeResize` (and friends)
We have `resize`, a function that can truncate by throwing away bits. We also have `checkedResize`, a function that does the same, but throws an error in simulation if the value of the argument does not fit in the result type. Ideally, we'd also have a total function that returns a `Maybe` instead.

### PR DESCRIPTION
**Pull request by jvnknvlgl**
[comment]: # (Thank you for contributing to Clash. Please fill out this template to describe your contribution and any outstanding tasks / questions.)
[comment]: # (External PRs will not automatically run on CI, this requires approval from a trusted contributer.)

Add `maybeResize` and `maybeTruncateB`. Resolves #2779.

[comment]: # (Add a line of the form "Fixes: #xxxx" for each related issue closed by this pull request, where #xxxx is an issue number.)

## Still TODO:

  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files

[comment]: # (Depending on the change, some of these points may not be necessary. If so, leave the boxes unchecked so it is easier for a reviewer to see what has been omitted.)

### CODE REVIEW

File: clash-prelude/src/Clash/Class/Resize.hs [hunk 1]
```diff
@@ -117,3 +120,57 @@ checkedTruncateB ::
...
+  a -> Maybe b
+maybeFromIntegral v =
+  if maybeIntegral (Proxy @b) v
+    then Just (fromIntegral v)
+    else Nothing
+
+-- | Like 'resize', but returns 'Nothing' if /f a/ is out of bounds for /f b/.
+-- Useful when you do not know /f a/ can be out of bounds, and would like to
+-- have your assumptions checked.
+maybeResize ::
```
Comment (rowanG077) [inline clash-prelude/src/Clash/Class/Resize.hs:153]: This will break for anything larger than native word size (64 bit). At least mention this limitation in the doc.

Comment (rowanG077) [reply clash-prelude/src/Clash/Class/Resize.hs:153]: Oh in fact you can do a short circuit on the type level if `BitSize a >= BitSize b` which means this function will be free in that case in hardware.

File: clash-prelude/src/Clash/Class/Resize.hs [hunk 2]
```diff
@@ -117,3 +120,57 @@ checkedTruncateB ::
...
+  Proxy b ->
+  a -> Bool
+maybeIntegral Proxy v =
+     toInteger v < toInteger (maxBound @b)
+  && toInteger v > toInteger (minBound @b)
+
+-- | Like 'fromIntegral', but returns 'Nothing' if /a/ is out of bounds for /b/.
+-- Useful when you do not know /a/ can be out of bounds, and would like to have
+-- your assumptions checked.
+
```
Comment (DigitalBrains1) [inline clash-prelude/src/Clash/Class/Resize.hs:137]: The Haddock for `maybeFromIntegral` is not generated, and that's probably because of this blank line. This blank line means that what follows the blank line is a regular Haskell comment rather than a Haddock comment. I would have expected the Haddock to still be associated with the function, but apparently it's cut off and lost?

TL;DR: Remove the blank line :-). And always check the Haddock output :-).

To view the Haddock, you can click on the _Details_ button next to the _ci/gitlab/gitlab.com_ CI test result, and then click on _haddock_ (in the column _test_), and then click the big _Browse_ button in the rightmost column of that webpage.

Or just generate Haddock on your own system of course, before you run CI.

File: clash-prelude/src/Clash/Class/Resize.hs [hunk 3]
```diff
@@ -117,3 +120,57 @@ checkedTruncateB ::
...
+  (Integral a, Integral b, Bounded b) =>
+  a -> Maybe b
+maybeFromIntegral v =
+  if maybeIntegral (Proxy @b) v
+    then Just (fromIntegral v)
+    else Nothing
+
+-- | Like 'resize', but returns 'Nothing' if /f a/ is out of bounds for /f b/.
+-- Useful when you do not know /f a/ can be out of bounds, and would like to
+-- have your assumptions checked.
```
Comment (DigitalBrains1) [inline clash-prelude/src/Clash/Class/Resize.hs:152]: I think all these `maybe` variants can be described much shorter:

> Like `resize`, but returns `Nothing` if _f a_ is out of bounds for _f b_.

The rest seems unnecessary, and also it mentions assumptions, but we don't go in with assumptions. This function is total; the assumption with the `checked` variant was that it would fit, but we want to error when that assumption is wrong. This is different.

Also, Martijn and I discussed that we'd best rephrase all these functions a bit, but that can be done in a followup PR. Or I can tack a commit onto this PR...

**Review by DigitalBrains1 — CHANGES_REQUESTED** [review]
I have some thoughts on the Haddock

**Comment by martijnbastiaan** [comment]
I've added HDL friendly versions and added tests @DigitalBrains1. Could you take another look? I've removed `maybeFromIntegral`, it seemed too big of a footgun. (Don't mind the CI changes, I'll struggle through all the old GHC versions at a later time..)

File: clash-prelude/tests/Clash/Tests/Resize.hs [hunk 4]
```diff
@@ -1,18 +1,37 @@
...
+
+import qualified Test.Tasty.QuickCheck as Q
+import qualified Hedgehog as H
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
 
 import qualified Clash.Class.Resize as Resize
-import Clash.Sized.Index
+import Clash.Promoted.Nat (SNat(..))
+import Clash.Prelude (Unsigned, Signed, BitVector, Index)
```
Comment (DigitalBrains1) [inline clash-prelude/tests/Clash/Tests/Resize.hs:30]: Did you intend to sort the imports like this? I got the sense you were separating non-qualified and qualified imports but here it's non-qualified again?

File: clash-prelude/tests/Clash/Tests/Resize.hs [hunk 5]
```diff
@@ -26,16 +45,146 @@ indexProp Proxy v =
...
+    maxFa = fromIntegral (maxBound @(f a)) :: Integer
+    minFb = fromIntegral (minBound @(f b)) :: Integer
+    maxFb = fromIntegral (maxBound @(f b)) :: Integer
+
+  H.footnote $ "minFa: " <> show minFa
+  H.footnote $ "maxFa: " <> show maxFa
+  H.footnote $ "minFb: " <> show minFb
+  H.footnote $ "maxFb: " <> show maxFb
+
+  input <- H.forAll $ Gen.integral (Range.linear minFa maxFa)
```
Comment (DigitalBrains1) [inline clash-prelude/tests/Clash/Tests/Resize.hs:83]: Wouldn't these be rather biased to zero? I.e., not producing interesting cases?
```suggestion
  input <- H.forAll $ Gen.integral (Range.constant minFa maxFa)
```

File: clash-prelude/tests/Clash/Tests/Resize.hs [hunk 6]
```diff
@@ -26,16 +45,146 @@ indexProp Proxy v =
...
+  let
+    minFa = fromIntegral (minBound @(f a)) :: Integer
+    maxFa = fromIntegral (maxBound @(f a)) :: Integer
+    minFab = fromIntegral (minBound @(f (a + b))) :: Integer
+    maxFab = fromIntegral (maxBound @(f (a + b))) :: Integer
+
+  H.footnote $ "minFa: " <> show minFa
+  H.footnote $ "maxFa: " <> show maxFa
+  H.footnote $ "minFab: " <> show minFab
+  H.footnote $ "maxFab: " <> show maxFab
```
Comment (DigitalBrains1) [inline clash-prelude/tests/Clash/Tests/Resize.hs:146]: This is a truly fabulous footnote!

(Also, thanks for adding footnotes, it's very nice in the case a counterexample is found)

File: clash-prelude/tests/Clash/Tests/Resize.hs [hunk 7]
```diff
@@ -26,16 +45,146 @@ indexProp Proxy v =
...
+
+  input <- H.forAll $ Gen.integral (Range.linear minFa maxFa)
```
Comment (DigitalBrains1) [inline clash-prelude/tests/Clash/Tests/Resize.hs:148]: Same comment about bias

```suggestion
  input <- H.forAll $ Gen.integral (Range.constant minFa maxFa)
```

File: clash-prelude/tests/Clash/Tests/Resize.hs [hunk 8]
```diff
@@ -49,5 +198,16 @@ tests = testGroup "Resize"
...
+    , testPropertyNamed "maybeResizeBitVectorProp" "maybeResizeBitVectorProp" maybeResizeBitVectorProp
+    , testPropertyNamed "maybeTruncateUnsignedProp" "maybeTruncateUnsignedProp" maybeTruncateUnsignedProp
+    , testPropertyNamed "maybeTruncateSignedProp" "maybeTruncateSignedProp" maybeTruncateSignedProp
+    , testPropertyNamed "maybeTruncateIndexProp" "maybeTruncateIndexProp" maybeTruncateIndexProp
+    , testPropertyNamed "maybeTruncateBitVectorProp" "maybeTruncateBitVectorProp" maybeTruncateBitVectorProp
     ]
   ]
+
+main :: IO ()
+main = defaultMain $ adjustOption (\_ -> HedgehogTestLimit (Just 1_000_000)) $ tests
```
Comment (DigitalBrains1) [inline clash-prelude/tests/Clash/Tests/Resize.hs:213]: Did you intend to include this in the PR or was it just for local testing?

**Review by DigitalBrains1 — CHANGES_REQUESTED** [review]
Thanks for going forward with this! I still have some thoughts.

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 5 above)
Comment (martijnbastiaan) [reply clash-prelude/tests/Clash/Tests/Resize.hs:83]: Yes, it would. I wasn't aware of `constant`! Actually, they both suck. They basically have zero chance of producing interesting edge cases.

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 4 above)
Comment (martijnbastiaan) [reply clash-prelude/tests/Clash/Tests/Resize.hs:30]: No rhyme or reason really :see_no_evil:. I'll fix it.

Comment (martijnbastiaan) [reply clash-prelude/tests/Clash/Tests/Resize.hs:30]: I want [`by-scope-then-qualified`](https://github.com/fourmolu/fourmolu/pull/403) and never manually sort imports again :heart_eyes:

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 5 above)
Comment (DigitalBrains1) [reply clash-prelude/tests/Clash/Tests/Resize.hs:83]: We know what the edge cases are, so we could produce primarily edge cases. [`Gen.frequency`](https://hackage.haskell.org/package/hedgehog-1.5/docs/Hedgehog-Gen.html#v:frequency).

You know, now that I look at it, the doc for `frequency` only says it shrinks towards the first generator. I really thought it also only considers earlier items on the list for small sizes. If I now look at the implementation, it would appear I am wrong about that.

File: clash-prelude/src/Clash/Class/Resize.hs [hunk 9]
```diff
@@ -117,3 +120,30 @@ checkedTruncateB ::
   f (a + b) -> f a
 checkedTruncateB v =
   checkIntegral (Proxy @(f a)) v `seq` truncateB v
+
+-- | Like 'resize', but returns 'Nothing' if /f a/ is out of bounds for /f b/.
```
[the remark this answers is not in the document]
Comment (DigitalBrains1) [reply clash-prelude/src/Clash/Class/Resize.hs:124]: Actually, this goes for the whole file, not just the addition.

Isn't this what I [referred to above](https://github.com/clash-lang/clash-compiler/pull/2780#discussion_r1705517838) by
>  Also, Martijn and I discussed that we'd best rephrase all these functions a bit, but that can be done in a followup PR. Or I can tack a commit onto this PR...

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 8 above)
Comment (martijnbastiaan) [reply clash-prelude/tests/Clash/Tests/Resize.hs:213]: Both! I'll make it less opinionated, but it's really nice for test modules to have `main`s. Then I can do:

```
./repld p:tests -W -T Clash.Tests.Resize.main
```

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 5 above)
Comment (martijnbastiaan) [reply clash-prelude/tests/Clash/Tests/Resize.hs]: Would you mind if I skip this (=> `frequency`. I've changed to `constant`.)? I don't want to spend much more effort. Or rather, I think the current situation is "good enough".

**Comment by martijnbastiaan** [comment]
@DigitalBrains1 I think I've applied all your suggestions.

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 5 above)
Comment (DigitalBrains1) [reply clash-prelude/tests/Clash/Tests/Resize.hs]: Go ahead. It's good enough for me with `constant`, I just continued on from your assertion that it still sucked :-D.

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 8 above)
Comment (DigitalBrains1) [reply clash-prelude/tests/Clash/Tests/Resize.hs]: ... _Or_ ...

We extend `repld` to allow passing [command line arguments](https://github.com/ndmitchell/ghcid?tab=readme-ov-file#how-do-i-run-pass-command-arguments-with---test) to `clash-prelude/tests/unittests.hs`, including arbitrrary Tasty options. You could do

```
./repld p:tests -W -T unittests -S Resize -A "--hedgehog-tests 1000000"
```
(S for _specific test_ and A for _additional arguments_)
and it would invoke
```
ghcid ... -T Main.main --setup ":set args -p Resize --hedgehog-tests 1000000"
```

File: clash-prelude/src/Clash/Class/Resize.hs [hunk 10]
```diff
@@ -117,3 +129,38 @@ checkedTruncateB ::
...
+  f a -> Maybe (f b)
+maybeResize v =
+#if MIN_VERSION_base(4,16,0)
+  case Proxy @a `cmpNat` Proxy @b of
+    GTI | v > resize (maxBound @(f b)) -> Nothing
+    GTI | v < resize (minBound @(f b)) -> Nothing
+    EQI -> Just v
+    _ -> Just (resize v)
+#else
+  case natToNum @a @Int `compare` natToNum @b @Int of
```
Comment (DigitalBrains1) [inline clash-prelude/src/Clash/Class/Resize.hs:149]: What about `Index (2^65)`? (Do we still do 32-bit machine compatibility? If so, what about `Index (2^33)`?)

I feel it would be better to use `Integer` or `Natural` here.

Comment (DigitalBrains1) [reply clash-prelude/src/Clash/Class/Resize.hs:149]:
```haskell
case natToNatural @a `compare` natToNatural @b of
```

Comment (martijnbastiaan) [reply clash-prelude/src/Clash/Class/Resize.hs:149]: Oh wow, I didn't even consider that. I guess the tests should pick from `[1..2^128]` too!

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 8 above)
Comment (martijnbastiaan) [reply clash-prelude/tests/Clash/Tests/Resize.hs]: Oh huh, I didn't know you could do that. Are you insisting on removing the `main`? I don't really mind it existing, even if the same thing can be achieved in another way.

**Comment by martijnbastiaan** [comment]
@jvnknvlgl I've removed `maybeFromIntegral`, as I think it is dangerous due to it having no good HDL interpretation. I wanted to explicitly mention it, because I feel we're close to merging this. If you feel like removing it was a mistake, please feel free to make a new PR/issue where we can discuss this.

**Review by DigitalBrains1 — APPROVED** [review]

**Comment by jvnknvlgl** [comment]
I agree with whatever changes you have made -  to be honest I completely forget about this PR anyway. Thanks for doing the work!

File: clash-prelude/tests/Clash/Tests/Resize.hs (hunk 8 above)
Comment (DigitalBrains1) [reply clash-prelude/tests/Clash/Tests/Resize.hs]: Oh no, I just wondered if there wasn't a "more nice" way to achieve the result, and this seemed like one. Go ahead and keep it as is.
